src/lib/y2storage/encryption_processes/tpm_fde_tools.rb

```diff
@@ -0,0 +1,142 @@
+# find current contact information at www.suse.com.
+
+require "yast2/execute"
+require "yast"
+require "y2storage/encryption_type"
+require "y2storage/encryption_processes/fde_tools"
+require "y2storage/encryption_processes/fde_tools_config"
+
+Yast.import "Mode"
+
+module Y2Storage
+  module EncryptionProcesses
+    # Encryption process that allow to setup device unlocking via the TPM2 chip
+    # based on the fde-tools shipped with (open)SUSE distributions
+    #
+    # The process is only valid for the system installation case, adding devices in an
+    # already installed system may imply different steps.
+    #
+    # This assumes all the devices use the same recovery password. That's a fde-tools
+    # requirement.
+    class TpmFdeTools < Base
+      # Options to add to the fourth column of crypttab for all involved devices
+      CRYPT_OPTIONS = ["x-initrd.attach"]
+      private_constant :CRYPT_OPTIONS
+
+      # Content of the third column of crypttab for all involved devices
+      KEY_FILE_NAME = "/.fde-virtual.key".freeze
+      private_constant :KEY_FILE_NAME
+
+      # Content of the third column of crypttab for all involved devices
+      #
+      # @return [String]
+      def self.key_file_name
+        KEY_FILE_NAME.dup
+      end
+
+      # Class methods
+      class << self
+        # List of all block devices configured by fde-tools during system installation
+        attr_accessor :devices
+      end
+
+      # Creates an encryption layer over the given block device
+      #
+      # @param blk_device [Y2Storage::BlkDevice]
+      # @param dm_name [String]
+      # @param label [String] optional LUKS2 label
+      #
+      # @return [Encryption]
+      def create_device(blk_device, dm_name, label: nil)
+        enc = super(blk_device, dm_name)
+        enc.label = label if label
+        enc.crypt_options |= CRYPT_OPTIONS
+        enc.key_file = self.class.key_file_name
+        enc.use_key_file_in_commit = false
+        # Maybe this configuration value affects only the slots added by fde-tools. In that
+        # case we could use any other PBKDF here. But being consistent with the PBKDF used
+        # in the fde-tools slots looks like a safer bet in this first implementation.
+        enc.pbkdf = FdeToolsConfig.instance.pbkd_function
+        enc
+      end
+
+      # @see Base#encryption_type
+      def encryption_type
+        EncryptionType::LUKS2
+      end
+
+      # @see Base#post_commit
+      #
+      # TODO: this implemented only for the installation case. In an installed system the procedure
+      # would be completely different and will likely include steps like configuring the fde-tools,
+      # calling "fdectl regenerate-key" and regenerating initrd and the bootloader configuration.
+      #
+      # @param device [Encryption] encryption that has just been created in the system
+      def post_commit(device)
+        return unless Yast::Mode.installation
+
+        self.class.devices ||= []
```

Comment:
BTW: why is `#devices` a class instance?

Reply:
Because there is a single list of devices for all the processes. We need to build a list with all the affected devices (which happens during `post_commit`) and then configure all those devices at once after chrooting. I will document that better.

Reply:
Improved the documentation to say basically the same than my previous comment.